### DISCUSSION TITLE
Separating product indexes based on clients in Pimcore

### DISCUSSION THREAD

**Discussion by MoliJina** [pimcore/pimcore#15384] [Q&A / Support]
Hello!

Let's say that we have multiple clients/tenants and we want to ensure that the product data for each client remains separate, including their respective product indexes.

To achieve this, I have set up workspaces/dataspaces where each client has their own dedicated folder structure. For example, there is a "client1" folder with client-specific folders and data, and a "client2" folder with its own set of folders and data.

However, when indexing process is triggered, I notice that products from each client's folder are being indexed in both the "client1-product-index" and "client2-product-index". This leads to the mixing of data from different clients in each index.

I would like to maintain data separation and ensure that each product is indexed in the correct client-specific product index. Could you please provide guidance on potential solutions or best practices to achieve this? 

I have found several threads saying Pimcore isn't really built for multi tenancy and in case of different sets of data usual solution is to just have separate Pimcore instances each one representing each client. Currently exploring/following this page from documentation: [Assortment Tenant Configuration](https://pimcore.com/docs/platform/Ecommerce_Framework/Index_Service/Product_Index_Configuration/Assortment_Tenant_Configuration/). Would be really helpful to at least know if it is doable and if this is the right track.

Pimcore version: 10.3.3

I appreciate any insights or recommendations you can provide to help us resolve this issue. Thank you for your assistance!

**Comment by kingjia90** [comment]
not an expert but my 2 cents: have you already tried to use the indexservice with `--tenant` option set?

![image](https://github.com/pimcore/pimcore/assets/6014195/c2d74759-6686-4ed0-acb9-fa25ce5d8fb5)
source: https://pimcore.com/docs/platform/Ecommerce_Framework/Index_Service/Product_Index_Configuration/Data_Architecture_and_Indexing_Process

**Reply by MoliJina** [reply]
Hey! Thanks for feedback! 

Yes, I've tried, if I specify that, it only affects the tenant I'm passing but it means I have to always use that and won't be able to just use `bin/console ecommerce:indexservice:bootstrap --create-or-update-index-structure`. I guess this would be fine but in this case(at least), indexing is happening when product is being updated/saved in Pimcore UI and it results in same thing(populating the updated product in all indexes) so it is not only the command that triggers it.

**Reply by kingjia90** [reply]
Ah, i see, so it's likely this https://github.com/pimcore/pimcore/blob/9f64dd623cb687dfb65f4844c53a3525be0a8a06/bundles/EcommerceFrameworkBundle/EventListener/IndexUpdateListener.php#L45
which is at the moment looping through all tenants
https://github.com/pimcore/pimcore/blob/9f64dd623cb687dfb65f4844c53a3525be0a8a06/bundles/EcommerceFrameworkBundle/IndexService/IndexService.php#L141-L143

**Reply by MoliJina** [reply]
Sorry, I gave wrong information previously. When specifying `--tenant` in the command, it only affects the respective tenant product index but it still indexes all other products from not-related dataspaces.(which is expected I suppose without somehow telling it which products to index and which not but that's what I'm trying to find out).

About your point, yes, was exploring there now. Just as an option, could restructure dataspaces so that dataspace and proper related tenant could be mapped through dataspace name, one option is to override `onObjectUpdate` and pass the name from path to `updateIndex` and then check against tenants there to decide which one to skip. I don't think it is a proper approach but I guess it is an option. Still unclear for the commands though.
